sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala

```diff
@@ -0,0 +1,194 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
+
+import com.google.common.io.ByteStreams
+
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.unsafe.Platform
+import org.apache.spark.unsafe.array.ByteArrayMethods
+import org.apache.spark.unsafe.memory.{TaskMemoryManager, MemoryBlock}
+import org.apache.spark.{TaskContext, SparkEnv}
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.io.CompressionCodec
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow
+
+/**
+ * ::Experimental::
+ *
+ * @since 1.6.0
+ */
+@Experimental
+object TungstenCache {
+  def cache(
+      child: SparkPlan,
+      compressionType: Option[String] = None,
+      blockSize: Int = 4 * 1000 * 1000): RDD[InternalRow] = {
+    val numFields = child.schema.length
+    val childRDD: RDD[UnsafeRow] = {
+      if (child.outputsUnsafeRows) {
+        child.execute().asInstanceOf[RDD[UnsafeRow]]
+      } else {
+        ConvertToUnsafe(child).execute().asInstanceOf[RDD[UnsafeRow]]
+      }
+    }
+    val cachedRDD: RDD[MemoryBlock] = childRDD.mapPartitions { rowIterator =>
+      // Note: this buffering is used to let us hold onto rows when rolling across block boundaries.
+      val bufferedRowIterator = rowIterator.buffered
+      val taskMemoryManager = TaskContext.get().taskMemoryManager()
+      val compressionCodec: Option[CompressionCodec] =
+        compressionType.map { t => CompressionCodec.createCodec(SparkEnv.get.conf, t) }
+      new Iterator[MemoryBlock] {
+        // NOTE: This assumes that size of every row < blockSize
```

Comment:
Actually, should this be -8 if we want to assume an end-of-block marker at the end of every page?